Initial bp-celery commit

monicating opened the pull request.

Commit:
Initial bp-celery commit

Commit:
Update README

monicating committed:
Update README.md

spladug commented on README.md:
```diff
@@ -1 +1,131 @@
-# baseplate-celery.py
```
These are really nice docs! Super thorough.

spladug commented on README.md:
```diff
@@ -1 +1,131 @@
+[app:celery-worker]
+
+# required: factory specifies the entry point for setting up Baseplate on the Celery worker
+factory = helloworld:make_celery_worker
```
where does `make_celery_worker` come from? I don't see it defined anywhere and the docs don't tell people how to make one

spladug commented on baseplate_celery/__init__.py:
```diff
@@ -0,0 +1,130 @@
+    def run_workers(self, app_config: config.RawConfig):
+        """ Starts the worker, sets up context for future use in workers """
+        # we use gevent pool so lets patch psycopg2 so it doesnt block
+        patch_psycopg()
```
Not every application uses PostgreSQL or psycopg2. Rather than putting this call here in the library and requiring every user of baseplate-celery to have psycopg2, could this just be left to the application's `make_celery_worker` function like we do for thrift/http?

spladug commented on baseplate_celery/__init__.py:
```diff
@@ -0,0 +1,130 @@
+    acks_late = True
+    track_started = True
+
+    _context = None
```
not being super familiar with celery, how do Task objects get instantiated? do we ever have to worry about concurrency here? if so might a task accidentally use the context from someone else?

spladug commented on baseplate_celery/__init__.py:
```diff
@@ -0,0 +1,130 @@
+        self._app_config = parse_celery_configs(baseplate._app_config)
+
+        secrets = secrets_store_from_config(baseplate._app_config, timeout=60)
+        session_factory = SQLAlchemySession(secrets).parse("", self._app_config.db)
```
what's this for? i thought we were using redis as the broker?

Tecuya replied:
This configures the result backend.  Deep originally set this up.  None of our current tasks use the result backend today. 

I think that `baseplate-celery` library should be totally agnostic about which result backend is used.  If users of this library want to configure a result backend they should be able to do it themselves by overriding something or setting config parameters.

Speaking of config parameters, [in the docs](https://github.com/reddit/baseplate-celery.py/pull/1/files/b30e5ec9c640c40ccd3b48b8c707d0375e6ff086#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R35) we instruct them to set a db result backend url in the config, as well as defining the a `credentials_secret` but I don't see that these `celery.db.*` values we reference are actually read anywhere.  Seems like this was a plan that never came to fruition?

spladug reviewed (changes requested):
Cool! Apologies for the delayed review. This looks really good. I have just a few questions about how it all fits together and how the app starts up. Do you have an example application maybe? That'd be cool to see all the bits in one place.

Naming wise: might I suggest `baseplate.py-celery` or just `baseplate-celery`? I doubt we'll have a baseplate-celery for other languages since it's a Python library so it's probably fine to go with the latter.

Thanks so much for doing this and I really appreciate all the details like docs and linters etc.

Tecuya commented:
Hey @monicating @spladug I'm Sean from reddit.  

Deep originally blazed the trail on this pattern, then I came in and modified it and got it running in CM.  

I'm enthusiastic to see this project.  We have plans to bring celery to other reddit services and having a common implementation would be best.

So, this pattern is in prod today, but it is rather specifically designed for use with `--pool gevent --concurrency 1` worker options.. like @spladug pointed out I'm not sure if the method we use to inject the context in to the task is safe for `concurrency` > 1.  

When I get some time I will do some testing and more research in to better patterns.  I'll leave some comments here as well.

monicating commented:
> I have just a few questions about how it all fits together and how the app starts up. Do you have an example application maybe? That'd be cool to see all the bits in one place.

Here is a PR with some changes on top of the work Deep/Sean had worked on for Campaign Management: https://github.snooguts.net/reddit/reddit-service-campaign-management/pull/3155

Tecuya commented:
We should incorporate changes from reddit/reddit-service-campaign-management/pull/3175/files

Commit:
Pull in task context changes

Commit:
Address comments

Commit:
README fixup

monicating commented:
Thanks @Tecuya for the improvements around adding the context on the task params! I've incorporated those changes here now and updated the README accordingly

spladug reviewed (approved):
looks good to me! excited to see this live.
